[WheelNext Summit] Summary (#22)
wheelnext/wheelnext: docs/summits/2025_03/summary.md

# 2025 WheelNext Community Summit - Summary

## **Slides and Resources:** [Link](site:summits/2025_03/slidedecks_and_resources)

## **Key Highlights**

- **Wheel Variants**
    - Critical for the scientific computing / machine and deep learning ecosystems.

>     - This ecosystem represent a very significant of python users, package downloads, etc.

Reviewer: ```suggestion
    - This ecosystem represent a very significant of Python users, package downloads, etc.
````

- **Index Group Priority** Priority-based resolution to mitigate supply chain risks.
- **Wheel Evolution** Modernize metadata and compression without breaking users.

> - **Governance** Push for a Packaging Council to unify standards.

Reviewer: ```suggestion
- **Governance** support formation of a Packaging Council, this will improve capacity of the community to review/absorb larger packaging PEPs.
```

    - Address security concerns (e.g. dependency confusion attacks)
    - No ability to "bypass build isolation" - difficulty to extend libraries depending on specific ABIs.
    - ZStandard compression would be a significant step forward

#### **Action Items**

- Keep being a driving force in the WheelNext Open-Source initiative.
- Keep driving forward the state of scientific computing, machine and deep learning.
- Commitment to support the Python Ecosystem at large and support the open source organizations behind it.

### **META/PyTorch** - Eli Uriegas

#### **PyTorch Foundation**:

- Aims to build a strong foundation for ML applications, encouraging community collaboration

#### **Challenges**:

- Space limitations on PyPI
- Currently need for a dedicated package index to support multiple hardware accelerators (we should do better!)
- Lack of features in packaging and installers (e.g. default extras, index priority, package variants, etc.)

#### **Future Improvements**:

- Variants proposal to improve packaging experience
- Index prioritization to manage multiple indices
- Default extras to enhance user experience

#### **Current Successes and Areas for Improvement**:

> - Adoption of ManyLinux 228 standard

Reviewer: ```suggestion
- Adoption of `manylinux_2_28` standard
```

- Managing CUDA environments and packaging for specialized libraries (e.g. BLAS, OpenMPI, etc.)
- Large binary sizes and lack of distro support

#### **Improvement Opportunities**:

- Adoption of new standards and best practices (e.g. WheelNext, [PEP 739](https://peps.python.org/pep-0739/))
- Enhanced accelerator support and binary size management
- Increased community collaboration and knowledge sharing

#### **Planned Contributions**:

- Active involvement in WheelNext development and PEP reviews
- Updates to [PyPackaging native](https://pypackaging-native.github.io/) website and completion of [PEP 725](https://peps.python.org/pep-0725/)
- Community building and outreach to improve packaging practices and standards

### **Red Hat** - Jeremy Eder

#### **Key Challenges**:

- Ensuring supply chain security and compliance (FIPS environments)
- Optimizing environments for business goals (cost, performance, stability)

#### **Improvement Opportunities**:

- Developing tools like [Fromager](https://github.com/python-wheel-build/fromager) for gap analysis and
community learning
- Centralizing capabilities based on community needs

#### **Growth Opportunities**:

> - Collaborating with community (Quansight, PEPs) to address impedance matching and flexibility

Reviewer: ```suggestion
- Collaborating with community (working with/through Quansight at the moment) to address impedance matching and flexibility
```

- Prioritizing customer feedback and business model alignment

#### **Lessons from Kubernetes**:

- Importance of customer feedback and business model alignment
- Evolution of features to address metadata problems and optimize deployments
- Scalability and adaptability: Kubernetes' success in scaling and adapting to changing requirements can inform
strategies for other projects and communities

### **AMD** - Jeff Daily

#### **Key Challenges**:

- Lack of ROCm packages for Python, leading to a significant gap in offerings
- Dependency management and compression algorithm issues
- User experience problems with package installations and environment conflicts

#### **Improvement Opportunities**:

- Developing variant packages for seamless user experience and smart defaults
- Increasing community presence and contributions from the packaging team

#### **Growth Opportunities**:

- Enhancing documentation and support for ROCm compatibility and usage
- Addressing compatibility queries and determining appropriate architecture combinations

#### **Future Directions**:

- Expanding ROCm package offerings and improving user experience through variant packages and community engagement

> - Collaborating with experts (e.g. Viya, Anoush) to improve documentation and compatibility support

Reviewer: are these names correct? I don't recognize them

Reviewer 2: This is like reading a bad transcript of what I said that was then paraphrased.  I called out @stellaraccident as I was speaking because she provided the bulk of our 1-page slide.  I called out Anush [@powderluv] as supporting our effort as part of our overall OSS commitment.  Let's just drop the name suggestions.
```suggestion
- Collaborating with experts to improve documentation and compatibility support
```

Author: @jeffdaily sorry Jeff - it was a lot of content to take note very fast :) Glad you caught that !

If your're good on the part for AMD, please approve the PR

### **Astral.sh** - Charlie Marsh

#### **Key Projects**:

- [Ruff](https://github.com/astral-sh/ruff): a static analysis tool chain with 45 million downloads per month
- [UV](https://github.com/astral-sh/uv): a Python package and project manager with 12.5% of PyPI download requests
- [Python Build Standalone](https://github.com/astral-sh/python-build-standalone): creating portable,
relocatable Python distributions

#### **Improvement Opportunities**:

- Developing declarative configuration and lock file consistency for simplified package installations
- Enhancing speed and user experience in Python packaging and installation
- Addressing challenges in hardware-accelerated dependencies and consistent metadata

#### **Growth Opportunities**:

> - Collaborating with the CPython community to upstream changes and simplify future development

Reviewer: ```suggestion
- Collaborating with the CPython community to upstream python-build-standalone changes and simplify future development
```

#### **Challenges and Concerns**:

- Variant handling and unclear user intent
- External channels and indices, including security implications
- Specifying non-Python dependencies for build and runtime

#### **Growth Opportunities**:

- Sharing experiences and aligning efforts with Python packaging
- Exploring package signing and verification methods
- Helping the community by sharing the learning accumulated with years of experience in variant related topics
in the `conda` ecosystem.

#### **Action Items**:

- Creating self-service wheel channels/indices for experimentation
- Improve interoperability between Conda and other Python packaging tools
- Investigate declaring non-Python dependencies in wheels
- Address security implications and user communication regarding channel priorities
- Facilitate self-service wheel channels on anaconda.org for experimentation

### **GPU Mode** - Mark Saroufim

#### **GPU Mode Community**:

- 15,000 kernel developers focused on ML systems
- Weekly YouTube lectures and working groups for project collaboration
- Kernel competition platform with 6,000 participants

#### **Challenges and Concerns**:

> - Interoperability issues between PyTorch and custom CPP code

Reviewer: ```suggestion
- Interoperability issues between PyTorch and custom C++ code
```

- Multi-backend Python packaging challenges
- JIT compilation time concerns for LLMs
- PyTorch binary size and CUDA toolkit installation issues

#### **Growth Opportunities**:

- Encouraging adoption of Wheelnext and growing its user base
- Addressing challenges and concerns to improve user experience

> - Providing resources and support for GPUmode community members

Reviewer: ```suggestion
- Providing resources and support for GPU MODE community members
```

#### **Action Items**:

> - Organizing talks with the GPU Mode community to prove the ground of ideas defined within wheelnext.

Reviewer: ```suggestion
- Organizing talks with the GPU MODE community to prove the ground of ideas defined within WheelNext.
```

> - Explore solutions for interoperability issues between PyTorch and custom CPP code

Reviewer: ```suggestion
- Explore solutions for interoperability issues between PyTorch and custom C++ code
```

    - **Communication and Feedback**:
        - Open PR for comments on the WheelNext repo.
        - Effective communication channels to inform the community about packaging standards and updates.

- **Action Items**
    - Follow up on Packaging Summit details at PyCon.
    - Draft governance model (similar to PEP 13).
    - Solicit feedback on initial council membership.
    - Participate on the discourse thread: <https://discuss.python.org/t/pep-772-packaging-governance-process/79724>

<!-- ------------------------------------------------------------------------------------------------------ -->

## 3. Wheel Variants

**Presenter:** Jonathan Dekhtiar

**Key Actors & Contributors:** NVIDIA, Quansight, META/PyTorch

**Details on the [Full Proposal](site:proposals/pepxxx_wheel_variant_support)**

### Takeaways

- Key points:
    - Python packaging limitations.
    - Wheel Variant concept in Python packaging.
    - Importance of feedback in development.

- Challenges:
    - Current packaging solutions are unsufficient to support the scientific computing / AI / ML / DL ecosystems.
    - Python package managers are totally hardware-unaware.

>     - No ability to describe a wheel beside extremely high level "tags" (python ABI, GCC version, OS, CPU Arch Family)

Reviewer: ```suggestion
   - No ability to describe a wheel beside extremely high level "tags" (Python ABI, GCC version, OS, CPU architecture family)
````

- Future considerations:
    - Future-proof design for variants.
    - Combination of arbitrary metadata.

>     - Shall not Impact on non-variant users.

Reviewer: ```suggestion
    - Shall not impact on non-variant users.
```

    - Feedback helps identify design issues.

- Importance in development:
    - Crucial in identifying small issues.
    - Improves design and functionality of the packaging system.

### Compatibility with Existing Installers

- Requirements:
    - Ensure compatibility with current installers.
    - Avoid breaking existing systems.

- Considerations:
    - New variant system should not break existing installers.
    - Compatibility with non-variant aware installers.

### Technical Proposal for Variants

- Definition:
    - Variants are wheels with additional metadata.
    - Metadata describes hardware and software configurations.

- Implementation:
    - Allows for arbitrary and combinable specifications.
    - Enables more specific and flexible packaging solutions.

<!-- ------------------------------------------------------------------------------------------------------ -->

## 4. Index Priority & Security

**Presenter:** Michael Sarahan

**Key Actors & Contributors:**  NVIDIA, META, PyTorch, Red Hat, Anaconda

**Details on the [Full Proposal](site:proposals/pep766_explicit_priority_choices)**

### Problem and Requirements

- Current limitations:

>     - PIP only supports the "latest and greatest" package.

Reviewer: ```suggestion
    - pip only supports the "latest and greatest" package.
```

    - No option to disable auto-advance behavior.

- Requirements:
    - Enable users to prioritize trust in a source or the latest package version.

>     - Maintain PIP’s assumption that all indexes are equal by merging them into one meta index.

Reviewer: ```suggestion
   - Maintain pip’s assumption that all indexes are equal by merging them into one meta index.
```

### Proposed Solution: Index Groups

- Concept:
    - Introduces "index groups" to allow different behaviors within and between groups.
- Behavior:

>     - Within a group: behavior remains as in current PIP.

Reviewer: ```suggestion
   - Within a group: behavior remains as in current pip.
```

- Importance of customer feedback and business model alignment
- Evolution of features to address metadata problems and optimize deployments
- Scalability and adaptability: Kubernetes' success in scaling and adapting to changing requirements can inform
strategies for other projects and communities

### **AMD** - Jeff Daily

#### **Key Challenges**:

- Lack of ROCm packages for Python, leading to a significant gap in offerings
- Dependency management and compression algorithm issues
- User experience problems with package installations and environment conflicts

#### **Improvement Opportunities**:

- Developing variant packages for seamless user experience and smart defaults
- Increasing community presence and contributions from the packaging team

#### **Growth Opportunities**:

- Enhancing documentation and support for ROCm compatibility and usage
- Addressing compatibility queries and determining appropriate architecture combinations

#### **Future Directions**:

- Expanding ROCm package offerings and improving user experience through variant packages and community engagement
- Collaborating with experts to improve documentation and compatibility support

> ### **Astral.sh** - Charlie Marsh

Reviewer 3: ```suggestion
### **Astral** - Charlie Marsh
```

#### **Key Projects**:

> - [Ruff](https://github.com/astral-sh/ruff): a static analysis tool chain with 45 million downloads per month

Reviewer 3: ```suggestion
- [Ruff](https://github.com/astral-sh/ruff): a static analysis toolchain with 45 million downloads per month
```

> - [UV](https://github.com/astral-sh/uv): a Python package and project manager with 12.5% of PyPI download requests

Reviewer 3: ```suggestion
- [uv](https://github.com/astral-sh/uv): a Python package and project manager with 12.5% of PyPI download requests
```

> - [Python Build Standalone](https://github.com/astral-sh/python-build-standalone): creating portable,

Reviewer 3: ```suggestion
- [python-build-standalone](https://github.com/astral-sh/python-build-standalone): creating portable,
```

- Interoperability issues between PyTorch and custom C++ code
- Multi-backend Python packaging challenges
- JIT compilation time concerns for LLMs
- PyTorch binary size and CUDA toolkit installation issues

#### **Growth Opportunities**:

- Encouraging adoption of WheelNext and growing its user base
- Addressing challenges and concerns to improve user experience
- Providing resources and support for GPU MODE community members

#### **Action Items**:

- Organizing talks with the GPU MODE community to prove the ground of ideas defined within WheelNext.
- Explore solutions for interoperability issues between PyTorch and custom C++ code
- Investigate ways to reduce JIT compilation time for LLMs
- Provide guidance on CUDA toolkit installation via Conda and address PyTorch binary size concerns

<!-- ------------------------------------------------------------------------------------------------------ -->

## 2. Python Packaging Governance

**Presenter:** Barry Warsaw

### Supporting PEP 772: Establish a Packaging Council

<https://peps.python.org/pep-0772/>

1. Establish formal governance over Python packaging standards, documentation, interoperability, and roadmaps.

> 2. Improve diversity of input and reduce the “bus factor”.

Reviewer 4: ```suggestion
2. Improve diversity of opinions and reduce the “bus factor”.
```
